Judge enhancements 04.2025

StoyanGoranov opened the pull request:
## Overview

This PR introduces the following enhancements: 

#### 1. According to FEI rules Disqualification can be done with some predetermined codes. Now we are adhering to FEI Rules
Highlights:
- now to add a custom reason to Disqualify the user must select **other** from the Multiselect and enter a reason manually
- validating when Multiselect is empty and also when **other** is selected but no Reason is provided

#### 2. Handouts page prompts the user to create Handouts manually when empty

#### 3. Making TimePickers used in CompetitionForm and ParticipationForm be editable with keyboard inputs
Highlights:
 - started as small tweak to prevent clumsiness;
 - expanded with validation of malformed time inputs in `CompetitionForm` and `ParticipationForm`;
 - new validation messages are localized;
 - also validating for ParticipationForms where Override properties are enabled but are left empty;
 - persisting Override checkboxes as checked, if Override values are set;
 - fixed an issue where `OverrideStartTime` cannot be set because `ToDateTimeOffset` returns a date in the past, if `TimeSpan` value is earlier than the current time because `ToDateTimeOffset` uses `DateTime.Today`

#### 4. Adding Main Menu header and localizin' it
- also removed obsolete Demos folder

StoyanGoranov committed:
add Disqualified Codes in EliminationPanel

this is according to official FEI rules

StoyanGoranov committed:
add HandoutsPage empty message

StoyanGoranov committed:
make TimePickers in Setup editable with keyboard inputs

StoyanGoranov committed:
validate a scenario where Time input in CompetitionForm is malformed

malformed input leads the TimeSpan field bound to the input to be null;
Consequently GetValueOrDefault(DateTime.Today.TimeOfDay) gets the current time and without any indication to the user the StartTime of a Competition is set to the current time.

StoyanGoranov committed:
fix a ParticipationForm issue where overriden values of Switches aren't
populated by the Form during editing

StoyanGoranov committed:
validate conditional inputs of ParticipationForm

StoyanGoranov committed:
fix casting TimeSpan ToDateTimeOffset and result is in the past

because TimeSpan value is earlier than the current time and
ToDateTimeOffset uses DateTime.Today

StoyanGoranov committed:
remove obsolete Localization Demos

StoyanGoranov committed:
add Main Menu header and localize it

achobanov commented on src/Judge/NTS.Judge.Blazor/Setup/EnduranceEvents/Participations/ParticipationForm.razor.cs:
```diff
@@ -1,19 +1,20 @@
 
 public partial class ParticipationForm
 {
+    static readonly PatternMask TIME_MASK = new("00:00");
```
Extract this time mask someplace and reuse it in CompetitionForm as well

achobanov replied:
I'm not sure we're going to keep the time mask approach as Mud's implementation behaves buggy, so having this centralized will help changing it, if we do so

StoyanGoranov replied:
Adding the time mask into a new NTS.Blazor project since it is going to be needed to share Components across Desktop, Mobile and Online environments.

achobanov commented on src/Judge/NTS.Judge.Blazor/Setup/EnduranceEvents/Participations/ParticipationFormModel.cs:
```diff
@@ -1,15 +1,37 @@
+            OverrideStartTime = value != null;
+            _startTimeOverride = value;
+        }
+    }
```
Why not simply assign `OverrideStartTime`'s value based on `StartTimeOverride`? Also naming is terrible.

achobanov commented on src/Judge/NTS.Judge.Blazor/Setup/EnduranceEvents/Participations/ParticipationFormModel.cs:
```diff
@@ -19,4 +41,16 @@ public void FromEntity(Participation participation)
         MaxSpeedOverride = participation.MaxSpeedOverride;
     }
+
+    public void ValidateConditionalInputs()
```
It shouldn't be `DomainException`. Those are supposed to be raised only for Domain model validation only.

achobanov commented on src/Judge/NTS.Judge.Blazor/Setup/EnduranceEvents/Competitions/CompetitionFormModel.cs:
```diff
@@ -58,6 +59,10 @@ public void FromEntity(Competition competition)
         var today = startDay.GetValueOrDefault(DateTime.Today);
+        if(startTime == null)
+        {
+            throw new DomainException(Null_or_malformed_string, Time_string);
```
Shouldn't be DomainException that's raised here.

achobanov commented on src/Judge/NTS.Judge/Features/Setup/Participations/SetupParticipationBehind.cs:
```diff
@@ -17,12 +17,14 @@ IEnumerable<ICrudReflection<Participation>> dependants
 
     protected override Participation UpdateEntity(ParticipationFormModel model)
     {
+        model.ValidateConditionalInputs();
```
Why not put the validation here instead of the Form model? Furthermore, what if the user toggled on, types a time and then toggles off? We shouldn't use the time, but we will

achobanov commented on src/NTS.Domain.Core/Aggregates/Participations/DqCode.cs:
```diff
@@ -0,0 +1,42 @@
+
+namespace NTS.Domain.Core.Aggregates.Participations;
+
+public enum DqCode
```
Lets rename these enums to `DisqualifyCode` and `FailToQualifyCode`.

achobanov commented on src/NTS.Domain.Core/Aggregates/Participations/DqCode.cs:
```diff
@@ -0,0 +1,42 @@
+{
+    /// <summary>
+    /// Underweight
+    /// </summary>
```
How is this summary useful?

achobanov commented on src/NTS.Domain.Core/Aggregates/Participations/Eliminated.cs:
```diff
@@ -28,12 +28,49 @@ public override string ToString()
+            throw new DomainException(
+                Please_provide_reason_to_eliminate_as__,
+                $"{Eliminated.DISQUALIFIED} {DqCode.other}"
+            );
```
Don't split stuff on new line manually please, Just leave it to the csharpier formatter

achobanov commented on src/Not/Not/Extensions/DateTimeOffsetExtension.cs:
```diff
@@ -9,9 +9,20 @@ public static DateTimeOffset ToDateTimeOffset(this DateTime time)
+        {
+            daysToAdd = 0;
+        }
+        var today = DateTime.Today.AddDays(daysToAdd);
```
This is not OK. Why are we moving this 1 day forward? These kind of manipulations need to be encapsulated if really necessary.

StoyanGoranov replied:
Sorry, this comment is unclear as to what to undertake given the explanation I have provided for the change.

achobanov committed:
Fix validation to Time field

github-actions[bot] committed:
Apply Csharpier code formatting

achobanov commented:
Can you also add `MinSpeedOverride` to Participation please? with the same toggle behavior or StartTime and MaxSpeedOverride

StoyanGoranov committed:
add NTS.Blazor project for common Components used across platforms

StoyanGoranov committed:
use new TIME_MASK_FORMAT constant from NTS.Blazor instead of hard coded strings

StoyanGoranov committed:
fix ParticipationFormModel and validate values when converting to Domain
Participation

StoyanGoranov committed:
rename DqCode to DisqualifyCode

StoyanGoranov committed:
rename FtqCode to FailToQualifyCode`

StoyanGoranov committed:
fix Elimiated.cs formatting

StoyanGoranov committed:
add MinSpeedOverride to Domain.Setup.Participation

StoyanGoranov committed:
add MinSpeedOverride to Setup ParticipationForm

StoyanGoranov committed:
localize MinSpeedOverride inputs

StoyanGoranov commented:
> Can you also add `MinSpeedOverride` to Participation please? with the same toggle behavior or StartTime and MaxSpeedOverride. Don't forget to include it in `UpcomingEventDocument` and it's mapping methods

Thanks for the reminder for `UpcomingEventDocument`

achobanov commented on src/NTS.Blazor/Component1.razor:
```diff
@@ -0,0 +1,3 @@
+﻿<div class="my-component">
```
Remove unnecessary stuff

achobanov commented on src/NTS.Storage/Documents/Archive/Models/EliminatedDocumentModel.cs:
```diff
@@ -14,8 +14,8 @@ public static EliminatedDocumentModel Create(Eliminated eliminated)
-        } 
-        else if(eliminated is Disqualified dq)
+        }
+        else if (eliminated is Disqualified dq)
```
`disqualified` over `dq`

StoyanGoranov committed:
Merge branch 'develop' into Judge-enhancements

StoyanGoranov committed:
reapply logic after merge with develop

StoyanGoranov committed:
remove unnecessary files from NTS.Blazor project

StoyanGoranov committed:
fix variable names in EliminatedDocumentModel

github-actions[bot] committed:
Apply Csharpier code formatting

StoyanGoranov committed:
move Mask constants from NTS.Blazor to NTS.Judge.Blazor.Constants

as at this point it is unlikely that Masks are going to be needed in
other projects; also NTS.Blazor is needed in Witness and will be added
in another branch/commit

github-actions[bot] committed:
Apply Csharpier code formatting

StoyanGoranov committed:
validate CompetitionFormModel @CompetitionBehind and combine time and
date

StoyanGoranov committed:
clean-up DateTimeOffsetExtension method

github-actions[bot] committed:
Apply Csharpier code formatting

StoyanGoranov committed:
beautify OverrideStartTime logic

StoyanGoranov committed:
Merge branch 'Judge-enhancements' of https://github.com/Not-Endurance/not-timing-system into Judge-enhancements

github-actions[bot] committed:
Apply Csharpier code formatting

achobanov committed:
Merge branch 'develop' into Judge-enhancements
